File: apps/web/tests/fixtures/portfolio.ts
import type {
  BrlDecomposition,
  EquityCurvePoint,
  FxObservation,
  KillSwitchDetail,
  PortfolioAnchor,
  PortfolioRiskState,
  PortfolioSummary,
} from "@/lib/api/portfolio-types";

/**
 * Fixtures for the T3.8b wallet screen. Shapes mirror the real Pydantic
 * schemas (`apps/api/hunter_api/schemas/{portfolio,portfolio_lists,risk}.py`)
 * -- every Decimal field is a plain decimal string, never a `number`.
 */

export function makeFxObservation(overrides: Partial<FxObservation> = {}): FxObservation {
  return {
    id: "fx-1",
    pair: "USDTBRL",
    source: "binance.spot.ticker",
    rate: "5.4000000000",
    observed_at: "2026-09-07T12:00:00Z",
    available_at: "2026-09-07T12:00:05Z",
    ...overrides,
  };
}

export function makeAnchor(overrides: Partial<PortfolioAnchor> = {}): PortfolioAnchor {
  return {
    portfolio_id: "wallet-1",
    origin_currency: "BRL",
    origin_amount: "100000.0000000000",
    operating_currency: "USDT",
    credited_amount: "18518.5185180000",
    rate: "5.4000000000",
    conversion_residual: "0.0000000200",
    rounding_policy: "floor_10dp_v1",
    anchored_at: "2026-09-01T13:00:00Z",
    fx_observation: makeFxObservation({ id: "fx-opening", observed_at: "2026-09-01T12:59:50Z" }),
    ...overrides,
  };
}

export function makeBrl(overrides: Partial<BrlDecomposition> = {}): BrlDecomposition {
  return {
    opening_brl: "100000.0000000000",
    operational_brl: "540.0000000000",
    currency_brl: "185.1851850000",
    equity_brl: "100725.1851850000",
    total_brl: "725.1851850000",
    opening_rate: "5.4000000000",
    current_rate: "5.4100000000",
    fx_observation: makeFxObservation(),
    ...overrides,
  };
}

export function makeRiskState(overrides: Partial<PortfolioRiskState> = {}): PortfolioRiskState {
  return {
    trading_day: "2026-09-07",
    trading_day_timezone: "America/Sao_Paulo",
    trading_day_start_utc: "2026-09-07T03:00:00Z",
    equity_day_start: "18600.0000000000",
    day_reference_observed_at: "2026-09-07T03:00:05Z",
    peak_equity: "18700.0000000000",
    peak_equity_observed_at: "2026-09-07T10:00:00Z",
    peak_sampling_interval_s: 60,
    daily_loss_pct: "-0.0050000000",
    drawdown_pct: "-0.0100000000",
    kill_switch: {
      effective: "ACTIVE",
      blocks_entries: false,
      scopes: { system: "ACTIVE", organization: "ACTIVE", portfolio: "ACTIVE" },
      reason: null,
      last_transition: null,
    },
    ...overrides,
  };
}

export function makeSummary(overrides: Partial<PortfolioSummary> = {}): PortfolioSummary {
  return {
    id: "wallet-1",
    organization_id: "org-1",
    workspace_id: "ws-1",
    name: "Carteira principal",
    type: "paper",
    status: "active",
    base_currency: "USDT",
    as_of: "2026-09-07T12:00:10Z",
    cash: "10000.0000000000",
    equity: "18650.5000000000",
    exposure_notional: "8650.5000000000",
    unrealized_pnl: "125.5000000000",
    realized_pnl_cum: "525.0000000000",
    open_position_count: 2,
    reserved_cash: "500.0000000000",
    reserved_notional: "1200.0000000000",
    reserved_risk: "46.6000000000",
    marks_complete: true,
    unavailable: [],
    brl: makeBrl(),
    brl_unavailable_reason: null,
    brl_unavailable_detail: null,
    risk_state: makeRiskState(),
    ...overrides,
  };
}

export function makeKillSwitch(overrides: Partial<KillSwitchDetail> = {}): KillSwitchDetail {
  return {
    portfolio_id: "wallet-1",
    effective: "ACTIVE",
    blocks_entries: false,
    scopes: { system: "ACTIVE", organization: "ACTIVE", portfolio: "ACTIVE" },
    reason: null,
    daily_reference: {
      trading_day: "2026-09-07",
      trading_day_timezone: "America/Sao_Paulo",
      trading_day_start_utc: "2026-09-07T03:00:00Z",
      equity_day_start: "18600.0000000000",
      observed_at: "2026-09-07T03:00:05Z",
      available: true,
    },
    peak: { equity: "18700.0000000000", observed_at: "2026-09-07T10:00:00Z", sampling_interval_s: 60 },
    last_transition: null,
    ...overrides,
  };
}

export function makeEquityPoint(overrides: Partial<EquityCurvePoint> = {}): EquityCurvePoint {
  return {
    ts: "2026-09-07T11:00:00Z",
    resolution: "1h",
    cash: "10000.0000000000",
    equity: "18650.5000000000",
    exposure_notional: "8650.5000000000",
    exposure_pct: "0.4640000000",
    unrealized_pnl: "125.5000000000",
    realized_pnl_cum: "525.0000000000",
    peak_equity: "18700.0000000000",
    drawdown_pct: "-0.0100000000",
    open_positions: 2,
    fx_observation_id: "fx-1",
    brl_equity: "100916.4550000000",
    brl_unavailable_reason: null,
    ...overrides,
  };
}

File: apps/api/hunter_api/schemas/portfolio.py
"""``/api/v1/orgs/{org_id}/portfolios`` — the wallet the T3.3 ledger already builds.

RISK_ENGINE.md §5 (kill switch), DATABASE.md §18.2 (the FX ledger) and the M3
joint decision, item 1 (BRL is "unavailable with a reason", never extrapolated).
This module is the *shape* of a read; the numbers themselves are assembled in
``hunter_api.services.portfolio_queries`` from ``hunter_core.portfolio`` and
``hunter_core.risk`` — nothing here recomputes a check or an attribution.
"""

from __future__ import annotations

import uuid
from datetime import date, datetime

from pydantic import BaseModel

from hunter_api.schemas.lab_common import DecimalStr
from hunter_api.schemas.risk import ScopeStatesOut, TransitionOut
from hunter_core.domain.enums import KillSwitchState, PortfolioStatus, PortfolioType


class PortfolioListItemOut(BaseModel):
    """One row of ``GET /api/v1/orgs/{org_id}/portfolios``."""

    id: uuid.UUID
    workspace_id: uuid.UUID
    name: str
    type: PortfolioType
    status: PortfolioStatus
    is_arena: bool
    base_currency: str
    created_at: datetime


class FxObservationOut(BaseModel):
    """One ``fx_observations`` row — global, immutable (DATABASE.md §18.2)."""

    id: uuid.UUID
    pair: str
    source: str
    rate: DecimalStr
    observed_at: datetime
    available_at: datetime


class AnchorOut(BaseModel):
    """``portfolio_currency_anchor`` — the opening conversion, written once."""

    portfolio_id: uuid.UUID
    origin_currency: str
    origin_amount: DecimalStr
    operating_currency: str
    credited_amount: DecimalStr
    rate: DecimalStr
    conversion_residual: DecimalStr
    rounding_policy: str
    anchored_at: datetime
    fx_observation: FxObservationOut


class BrlDecompositionOut(BaseModel):
    """The operational/exchange split — ``hunter_core.portfolio.attribution``.

    ``operational_brl`` and ``currency_brl`` sum to ``total_brl``; the
    convention (attribution at the *opening* rate) is declared, not universal,
    and this is why the screen names ``opening_rate``/``current_rate`` instead
    of hiding them behind the totals.
    """

    opening_brl: DecimalStr
    operational_brl: DecimalStr
    currency_brl: DecimalStr
    equity_brl: DecimalStr
    total_brl: DecimalStr
    opening_rate: DecimalStr
    current_rate: DecimalStr
    fx_observation: FxObservationOut


class KillSwitchSummaryOut(BaseModel):
    """The kill switch, embedded in the portfolio summary.

    Mirrors ``hunter_api.schemas.risk.KillSwitchOut`` minus the daily
    reference/peak (reported once, in ``PortfolioRiskStateOut``, not twice).
    """

    effective: KillSwitchState
    blocks_entries: bool
    scopes: ScopeStatesOut
    reason: str | None
    last_transition: TransitionOut | None


class PortfolioRiskStateOut(BaseModel):
    """``portfolio_risk_state`` plus the engine's own read of it.

    ``daily_loss_pct``/``drawdown_pct`` are ``None`` exactly when
    ``hunter_core.portfolio.state.build_portfolio_state`` could not rebuild the
    daily reference (a restart before the first evaluation of the day) — never
    a silent zero (RISK_ENGINE.md §5, the "daily_loss PASSED value=0" defect
    this contract was written to close).
    """

    trading_day: date | None
    trading_day_timezone: str
    trading_day_start_utc: datetime | None
    equity_day_start: DecimalStr | None
    day_reference_observed_at: datetime | None
    peak_equity: DecimalStr
    peak_equity_observed_at: datetime
    peak_sampling_interval_s: int
    daily_loss_pct: DecimalStr | None
    drawdown_pct: DecimalStr | None
    kill_switch: KillSwitchSummaryOut


class PortfolioSummaryOut(BaseModel):
    """``GET /api/v1/orgs/{org_id}/portfolios/{portfolio_id}``.

    ``unavailable`` names *why*, in the words ``PortfolioStateBuild`` uses
    (``"marks"``, ``"market_identity"``, ``"daily_reference"``,
    ``"daily_decomposition"``) — never a silent gap in the numbers above it.
    """

    id: uuid.UUID
    organization_id: uuid.UUID
    workspace_id: uuid.UUID
    name: str
    type: PortfolioType
    status: PortfolioStatus
    base_currency: str
    as_of: datetime
    cash: DecimalStr
    equity: DecimalStr
    exposure_notional: DecimalStr
    unrealized_pnl: DecimalStr
    realized_pnl_cum: DecimalStr
    open_position_count: int
    reserved_cash: DecimalStr
    reserved_notional: DecimalStr
    reserved_risk: DecimalStr
    marks_complete: bool
    unavailable: list[str]
    brl: BrlDecompositionOut | None
    brl_unavailable_reason: str | None
    brl_unavailable_detail: str | None
    risk_state: PortfolioRiskStateOut
